File: csharp/SortQuick.cs
public class SortQuick : SortBenchmark
{
    protected override int[] Test()
    {
        int[] arr = new int[_data.Length];
        Array.Copy(_data, arr, _data.Length);
        QuickSort(arr, 0, arr.Length - 1);
        return arr;
    }

    private void QuickSort(int[] arr, int low, int high)
    {
        if (low >= high) return;

        int pivot = arr[(low + high) / 2];
        int i = low, j = high;

        while (i <= j)
        {
            while (arr[i] < pivot) i++;
            while (arr[j] > pivot) j--;

            if (i <= j)
            {
                (arr[i], arr[j]) = (arr[j], arr[i]);
                i++;
                j--;
            }
        }

        QuickSort(arr, low, j);
        QuickSort(arr, i, high);
    }
}

File: swift/Sources/benchmarks/SortQuick.swift
import Foundation

final class SortQuick: SortBenchmark {
  override init() {
    super.init()
  }

  private func quickSort(_ arr: inout [Int], low: Int, high: Int) {
    if low >= high { return }

    let pivot = arr[(low + high) / 2]
    var i = low
    var j = high

    while i <= j {
      while arr[i] < pivot { i += 1 }
      while arr[j] > pivot { j -= 1 }
      if i <= j {
        arr.swapAt(i, j)
        i += 1
        j -= 1
      }
    }

    quickSort(&arr, low: low, high: j)
    quickSort(&arr, low: i, high: high)
  }

  override func test() -> [Int] {
    var arr = data
    quickSort(&arr, low: 0, high: arr.count - 1)
    return arr
  }

  override func name() -> String {
    return "Sort::Quick"
  }
}
